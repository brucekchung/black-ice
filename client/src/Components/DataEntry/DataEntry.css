.DataEntry {
  text-align: center;
  font-weight: 300;
  /*margin-top: 50px;*/
}

.DataEntry h3 {
  font-size: 1.5rem;
  font-weight: 900;
  margin-top: 50px;
  margin-bottom: 10px;
}

.DataEntry h4 {
  font-size: 1.25rem;
  margin-bottom: 30px;
}

.DataEntry h5 {
  font-size: 1.25rem;
  margin-bottom: 20px;
}

.import-button {
  position: absolute;
  right: 46px;
  top: 150px;
}

.choose-location,
.add-location {
  margin-bottom: 20px;
}

.submit-button {
  margin: 70px 0 30px 0;
}