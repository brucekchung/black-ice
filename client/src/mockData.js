export const allData = [{
    sample_id: '1',
    name: 'Base Camp',
    date_collected: '08/19/18',
    reflectance: '89',
    wavelength: '20',
    location_id: '1',
    altitude: '4000m',
    lat: '1',
    lng: '-1',
    region: 'Ishinca',
    country: 'Peru'
  }, {
    sample_id: '2',
    name: 'Camp 1',
    date_collected: '08/20/18',
    reflectance: '70',
    wavelength: '110',
    location_id: '2',
    altitude: '5000m',
    lat: '2',
    lng: '-2',
    region: 'Pisco',
    country: 'Peru'       
}]

export const locationsOnly = [{
    id: '1',
    name: 'Base Camp',
    lat: '1',
    lng: '-1',
    alt: '4000m',
    region: 'Ishinca',
    country: 'Argentina'    
  }, {
    id: '2',
    name: 'Camp 1',
    lat: '2',
    lng: '-2',
    alt: '5000m',
    region: 'Pisco',
    country: 'Peru'
  }, {
    id: '3',
    name: 'Camp 2',
    lat: '1',
    lng: '-2',
    alt: '6000m',
    region: 'Ishinca',
    country: 'Argentina'
}]

export const dataOnly = [{
    id: '1',
    name: 'WP 1',
    date_collected: '08/19/18',
    reflectance: '89',
    wavelength: '20',
    location_id: '1'
  }, {
    id: '2',
    name: 'WP 22',
    date_collected: '08/20/18',
    reflectance: '70',
    wavelength: '110',
    location_id: '2'
}]