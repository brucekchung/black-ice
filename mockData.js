export const allData = [{
    id: '1',
    name: 'Base Camp',
    date_collected: '08/19/18',
    reflectance: '89%',
    wavelength: '20um',
    altitude: '4000m',
    latlng: '1, -1',
    region: 'Ishinca'
    country: 'test'
  }, {
    id: '2',
    name: 'Camp 1',
    date_collected: '08/20/18',
    reflectance: '70%',
    wavelength: '110um',
    altitude: '5000m',
    latlng: '2, -2',
    region: 'Pisco'
    country: 'test2'       
}]

export const allLocations = [{
    id: '1',
    name: 'Base Camp',
    lat: '1',
    lng: '-1',
    altitude: '4000m',
    region: 'Ishinca',
    country: 'Peru'    
  }, {
    id: '2',
    name: 'Camp 1',
    lat: '2',
    lng: '-2',
    altitude: '5000m',
    region: 'Pisco',
    country: 'Peru'
}]

export const data = [{
    id: '1',
    name: 'WP 1',
    date_collected: '08/19/18',
    reflectance: '89%',
    wavelength: '20um',
    location_id: '1'
  }, {
    id: '2',
    name: 'WP 22',
    date_collected: '08/20/18',
    reflectance: '70%',
    wavelength: '110um',
    location_id: '2'
}]