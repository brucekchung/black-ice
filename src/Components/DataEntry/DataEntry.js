import React, { Component } from 'react'
import './DataEntry.css'
import Nav from '../Nav/Nav'
import DropDown from '../LocationForm/LocationForm'
import DataForm from '../DataForm/DataForm'

class DataEntry extends Component {
  constructor() {
    super()
    this.state = {
      options: {
        countries: ['Peru', 'Argentina'],
        coordinates: ['1, -1'],
      },
      location: {
        country: null,
        coordinates: null
      },
      data: [{
          name: '',
          date_collected: '',
          reflectance: '',
          wavelength: ''
      }]
    }
  }

  handleChange = (e) => {
    const name = e.target.name
    const value = e.target.options[e.target.selectedIndex].value

    this.setState({ 
      location: { [name]: value }
    })
  }

  handleClick = (e) => {
    e.preventDefault();
    e.target.innerText === 'Add Data' ? this.addDataSet() : this.removeDataSet()
  }

  updateData = (e) => {
    const name = e.target.name
    const value = e.target.value
    const index = e.target.parentNode.className
    const data = [...this.state.data]
    
    data[index][name] = value
    this.setState({ data })
  }

  addDataSet = () => {
    const newSet = {
      name: '',
      date_collected: '',
      reflectance: '',
      wavelength: ''
    }
    const newDataArray = [...this.state.data, newSet]

    this.setState({ data: newDataArray })
  }

  removeDataSet = () => {
    if (this.state.data.length > 1) {
      const allItems = [...this.state.data]

      allItems.pop();
      this.setState({ data: allItems })
    }
  }

  render() {
    return (
      <div>
        <Nav />
        <h2>Add Data</h2>
        <button>Import CSV</button>
        <form>
          <h3>Location</h3>
          <DropDown name='country'
                    options={ this.state.options.countries }
                    handleChange={ this.handleChange } />
          <DropDown name='coordinates'
                    options={ this.state.options.coordinates }
                    handleChange={ this.handleChange } />
          <h3>Data</h3>
          <DataForm data={ this.state.data }
                    handleChange={ this.updateData }
                    handleClick={ this.handleClick } />
          <button type="submit">Save</button>
        </form>
      </div>
    )
  }
}

export default DataEntry
