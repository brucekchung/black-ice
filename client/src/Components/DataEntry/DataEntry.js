import React, { Component } from 'react'
import './DataEntry.css'
import Nav from '../Nav/Nav'
import DropDown from '../LocationForm/LocationForm'
import DataForm from '../DataForm/DataForm'
import { locationsOnly } from '../../mockData'

class DataEntry extends Component {
  constructor() {
    super()
    this.state = {
      options: {
        countries: [],
        coordinates: [],
        regions: []
      },
      location: {
        country: null,
        coordinates: null,
        region: null
      },
      data: [{
          name: '',
          date_collected: '',
          reflectance: '',
          wavelength: ''
      }]
    }
  }

  componentDidMount = async () => {
    const url = '/api/v1/locations'
    // const locations = await apiCall(url);
    const countries = this.sortOptionsInfo(locationsOnly, 'country')
    const regions = this.sortOptionsInfo(locationsOnly, 'region')
    const coordinates = this.sortLatLng(locationsOnly);

    this.setState({ 
      options: {
        countries,
        regions,
        coordinates
      }
    })
  }

  sortOptionsInfo = (locations, type) => {
    return locations.reduce((acc, location) => {
      if(acc.includes(location[type])) {
        return acc
      }

      return [...acc, location[type]]
    }, [])
  }

  sortLatLng = locations => {
    return locations.reduce((acc, location) => {
      const latlng = `${location.lat}, ${location.lng}`

      if(acc.includes(latlng)) {
        return acc
      }

      return [...acc, latlng]
    }, [])
  }

  handleChange = (e) => {
    const name = e.target.name
    const value = e.target.options[e.target.selectedIndex].value

    this.setState({ 
      location: { [name]: value }
    })
  }

  handleClick = (e) => {
    e.preventDefault();
    e.target.innerText === 'Add Data' ? this.addDataSet() : this.removeDataSet()
  }

  updateData = (e) => {
    const name = e.target.name
    const value = e.target.value
    const index = e.target.parentNode.className
    const data = [...this.state.data]
    
    data[index][name] = value
    this.setState({ data })
  }

  addDataSet = () => {
    const newSet = {
      name: '',
      date_collected: '',
      reflectance: '',
      wavelength: ''
    }
    const newDataArray = [...this.state.data, newSet]

    this.setState({ data: newDataArray })
  }

  removeDataSet = () => {
    if (this.state.data.length > 1) {
      const allItems = [...this.state.data]

      allItems.pop()
      this.setState({ data: allItems })
    }
  }

  handleSubmit = e => {
    e.preventDefault()

    // const dataToSend = {
    //   location_id,
    //   data: this.state.data,
    // }
    // const url = ''
    // const init = { 
    //   method: 'POST',
    //   headers: {
    //     'Content-Type': 'application/json'
    //   },
    //   body: JSON.stringify(dataToSend)
    // }

    // apiCall(url, init)    
  }

  render() {
    return (
      <div>
        <Nav />
        <h2>Add Data</h2>
        <button>Import CSV</button>
        <form onSubmit={ this.handleSubmit }>
          <h3>Location</h3>
          <DropDown name='country'
                    options={ this.state.options.countries }
                    handleChange={ this.handleChange } />
          <DropDown name='region'
                    options={ this.state.options.regions }
                    handleChange={ this.handleChange } />
          <DropDown name='coordinates'
                    options={ this.state.options.coordinates }
                    handleChange={ this.handleChange } />
          <h3>Data</h3>
          <DataForm data={ this.state.data }
                    handleChange={ this.updateData }
                    handleClick={ this.handleClick } />
          <button type="submit">Save</button>
        </form>
      </div>
    )
  }
}

export default DataEntry
