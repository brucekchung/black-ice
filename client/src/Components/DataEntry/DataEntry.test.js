/* eslint-disable */
import React from 'react'
import DataEntry from './DataEntry'
import { shallow } from 'enzyme'
import { locationsOnly } from '../../mockData'

describe('DataEntry', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<DataEntry />, { disableLifecycleMethods: true })
  })

  it('should match snapshot upon load', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should match snapshot with locations', () => {
    wrapper.setState({ allLocations: locationsOnly })

    expect(wrapper).toMatchSnapshot()
  })

  it('should have a default state', () => {
    const expected = {
      allLocations: [],
      selectedLocation: {
        name: '',
        country: '',
        coordinates: '',
        region: '',
        lat: '',
        lng: '',
        alt: ''
      },
      addLocation: false,
      sampleData: [{
          name: '',
          date_collected: '',
          reflectance: '',
          wavelength: ''
      }]
    }

    expect(wrapper.state()).toEqual(expected)
  })

  it('should update the state of addLocation when a user wants to add a new location', () => {
    const e = {
      target: {
        innerText: 'Add a New Location'
      }
    }

    wrapper.instance().changeLocationForm(e);
    expect(wrapper.state('addLocation')).toEqual(true)
  })

  it('should not render duplicate countries and regions in the drop down', () => {
    const locations = locationsOnly
    const expected = ['Argentina', 'Peru']
    const options = wrapper.instance().sortOptionsInfo(locations, 'country')

    expect(options).toEqual(expected)
  })

  it('should format the coordinates correctly', () => {
    const locations = [{ lat: '1', lng: '-1'}]
    const expected = ['1, -1']
    const coords = wrapper.instance().sortLatLng(locations)

    expect(coords).toEqual(expected)

  })

  it('should update the state of selectedLocation when an input is changed', () => {
    const e = {
      target: {
        name: 'country',
        options: [{ value: 'Peru'}],
        selectedIndex: 0
      }
    }
    const expected = {
      country: 'Peru',
      alt: '',
      lat: '',
      lng: '',
      name: '',
      coordinates: '',
      region: ''
    }

    wrapper.instance().handleChange(e)
    expect(wrapper.state('selectedLocation')).toEqual(expected)
  })

  it('should add a data form when the add data button is clicked', () => {
    const e = {
      target: {
        innerText: 'Add Data'
      },
      preventDefault: jest.fn()
    }

    expect(wrapper.state('sampleData').length).toEqual(1)
    wrapper.instance().handleClick(e)
    expect(wrapper.state('sampleData').length).toEqual(2)
  })

  it.skip('should update each data object in state when the associated input is changed', () => {
    const e = {
      target: {
        name: 'name',
        value: 'TEST',
        parentNode: {
          className: 0
        }
      }
    }
    const expected = [{
      name: 'TEST',
      date_collected: '',
      reflectance: '',
      wavelength: ''
    }]

    wrapper.instance().updateData(e)
    expect(wrapper.state('sampleData')).toEqual(expected)
  })

  it('should add another data object to state', () => {
    expect(wrapper.state('sampleData').length).toEqual(1)
    wrapper.instance().addDataSet()
    expect(wrapper.state('sampleData').length).toEqual(2)
  })

  it('should remove the last data form when the remove button is clicked', () => {
    wrapper.instance().addDataSet()
    expect(wrapper.state('sampleData').length).toEqual(2)

    wrapper.instance().removeDataSet()
    expect(wrapper.state('sampleData').length).toEqual(1)
  })

  it('should call fetch with the correct parameters when POSTing data', () => {
    window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
      json: () => Promise.resolve({})
    }))
    const locations = {
      type: 'locations',
      payload: locationsOnly
    }
    const url = '/api/v1/locations'
    const init = {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(locationsOnly)
    }

    wrapper.instance().sendData(locations)
    expect(window.fetch).toHaveBeenCalledWith(url, init)

  })

  it('should call fetch with the correct parameters when GETting location id', () => {
    wrapper.setState({
      selectedLocation: {
        name: '',
        country: '',
        coordinates: '1, -1',
        region: '',
        lat: '',
        lng: '',
        alt: ''
      }
    })
    window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
      json: () => Promise.resolve([{ id: 1 }])
    }))

    const id = wrapper.instance().getLocationId()
    const url = '/api/v1/locations?lat=1&lng=-1'
    const init = null
    
    expect(window.fetch).toHaveBeenCalledWith(url, init)
    // expect(id).toEqual(1)
  })

  it('should reset the state', () => {
    wrapper.setState({
      selectedLocation: {
        name: '',
        country: '',
        coordinates: '1, -1',
        region: '',
        lat: '',
        lng: '',
        alt: ''
      }
    })

    const expected = {
      allLocations: [],
      selectedLocation: {
        name: '',
        country: '',
        coordinates: '',
        region: '',
        lat: '',
        lng: '',
        alt: ''
      },
      addLocation: false,
      sampleData: [{
          name: '',
          date_collected: '',
          reflectance: '',
          wavelength: ''
      }]
    }

    wrapper.instance().resetState()
    expect(wrapper.state()).toEqual(expected)
  })

  it('should call fetch and reset state', () => {
    window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
      json: () => Promise.resolve([{ id: 1 }])
    }))
    const expected = {
      allLocations: [],
      selectedLocation: {
        name: '',
        country: '',
        coordinates: '',
        region: '',
        lat: '',
        lng: '',
        alt: ''
      },
      addLocation: false,
      sampleData: [{
          name: '',
          date_collected: '',
          reflectance: '',
          wavelength: ''
      }]
    }

    wrapper.instance().submit()
    expect(window.fetch).toHaveBeenCalled()
    expect(wrapper.state()).toEqual(expected)
  })

  it('should POST a new location', () => {
    window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
      json: () => Promise.resolve({})
    }))

    const payload = {
      country: '',
      region: '',
      lat: '',
      lng: '',
      name: '',
      alt: ''
    }
    const url = '/api/v1/locations'
    const init = {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(payload)
    }

    wrapper.instance().submitLocation()
    expect(window.fetch).toHaveBeenCalledWith(url, init)
  })

  it('should POST a location and data, then reset state', () => {
    window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
      json: () => Promise.resolve({})
    }))
    const expected = {
      allLocations: [],
      selectedLocation: {
        name: '',
        country: '',
        coordinates: '',
        region: '',
        lat: '',
        lng: '',
        alt: ''
      },
      addLocation: false,
      sampleData: [{
          name: '',
          date_collected: '',
          reflectance: '',
          wavelength: ''
      }]
    }

    wrapper.instance().submitWithLocation()
    expect(window.fetch).toHaveBeenCalled()
    expect(wrapper.state()).toEqual(expected)
  })

  it('should call the correct function based on the what location form is chosen', () => {
    wrapper.instance().submit = jest.fn()
    wrapper.instance().submitWithLocation = jest.fn()
    const e = {
      preventDefault: jest.fn()
    }

    wrapper.instance().handleSubmit(e)
    expect(wrapper.instance().submit).toHaveBeenCalled()

    wrapper.setState({ addLocation: true })
    wrapper.instance().handleSubmit(e)
    expect(wrapper.instance().submitWithLocation).toHaveBeenCalled()
  })

  it('should render the correct options for the remaining location drop downs', () => {
    wrapper.instance().sortOptionsInfo = jest.fn()
    wrapper.instance().sortLatLng = jest.fn()

    wrapper.instance().availableOptions('region')
    expect(wrapper.instance().sortOptionsInfo).toHaveBeenCalled()

    wrapper.instance().availableOptions('coordinates')
    expect(wrapper.instance().sortLatLng).toHaveBeenCalled()
  })
})
