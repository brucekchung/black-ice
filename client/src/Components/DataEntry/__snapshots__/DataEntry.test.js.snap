// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataEntry should match snapshot upon load 1`] = `undefined`;
